File: src/prod/src/data/tstore/PropertyChunkMetadata.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"
using namespace ktl;
using namespace Data::TStore;
using namespace Data::Utilities;

const ULONG32 PropertyChunkMetadata::Size;

PropertyChunkMetadata::PropertyChunkMetadata(__in ULONG32 blockSize)
   :blockSize_(blockSize)
{
}

PropertyChunkMetadata::~PropertyChunkMetadata()
{
}

PropertyChunkMetadata PropertyChunkMetadata::Read(__in BinaryReader& reader)
{
   ByteAlignedReaderWriterHelper::ThrowIfNotAligned(reader.Position);

   ULONG32 currentBlockSize = 0;
   reader.Read(currentBlockSize);
   ByteAlignedReaderWriterHelper::ReadPaddingUntilAligned(reader);

   PropertyChunkMetadata propertyChunkMetadata(currentBlockSize);

   return propertyChunkMetadata;
}

void PropertyChunkMetadata::Write(__in BinaryWriter& writer)
{
   ByteAlignedReaderWriterHelper::AssertIfNotAligned(writer.Position);

   writer.Write(blockSize_);
   ByteAlignedReaderWriterHelper::WritePaddingUntilAligned(writer);

   ByteAlignedReaderWriterHelper::AssertIfNotAligned(writer.Position);
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/ReplicatedStore/Disk/PropertyChunkMetadata.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Store
{
    using Microsoft.ServiceFabric.Replicator;

    /// <summary>
    /// Metadata about the property block written at the begining of the block.
    /// Can be considered as a header.
    /// </summary>
    /// <devnote>
    /// MCoskun: Note that KeyBlockMetadata.BlockSize includes KeyBlockMetadata and KeyBlock.
    /// It does not include the checksum
    /// </devnote>
    /// <devnote>
    /// MCoskun: Note that checksum includes the KeyBlockMetadata and KeyBlock.
    /// </devnote>
    /// <devnote>
    /// TODO: We could use GCHandle and StructureToPtr to do copy free serialization and de-serialization.
    /// </devnote>
    internal struct PropertyChunkMetadata
    {
        /// <summary>
        /// int for size field and 4 bytes of reserved space.
        /// </summary>
        public const int Size = sizeof(int) + 4;

        public readonly int BlockSize;

        public PropertyChunkMetadata(int blockSize)
        {
            this.BlockSize = blockSize;
        }

        public static PropertyChunkMetadata Read(InMemoryBinaryReader reader)
        {
            reader.ThrowIfNotAligned();

            var currentBlockSize = reader.ReadInt32();
            reader.ReadPaddingUntilAligned(true);

            reader.ThrowIfNotAligned();

            return new PropertyChunkMetadata(currentBlockSize);
        }

        public void Write(InMemoryBinaryWriter writer)
        {
            writer.AssertIfNotAligned();

            writer.Write(this.BlockSize);
            writer.WritePaddingUntilAligned();

            writer.AssertIfNotAligned();
        }
    }
}